{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To download all fasta files and compile from source, you can run the provided script with the input file from the GTDB project. Next, use the `--add-to-library` and `--build `functions in Kraken to format the database. Example commands are below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Error 429 too many requests. NCBI is receiving too many general requests, but you can carve out a special place for yourself by getting an NCBI API key. Log into your NCBI profile and copy your key. Then, add it to your environment like so: `export NCBI_API_KEY=1fe2...`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Kraken2 is a k-mer-based classifier used for species classification of high-throughput sequencing data. When building the Kraken database, one can choose to include sequences from the NCBI non-redundant nucleotide database (NT) to increase the accuracy and coverage of classification.\n",
    "\n",
    "The NT database contains DNA and RNA sequences from various species and sources, including genes, transcripts, non-coding RNAs, genomes, viruses, plasmids, and more. These sequences cover a wide range of biological diversity from different organisms, pathogens, and host cells.\n",
    "\n",
    "Including the NT database in the Kraken database may be helpful if you want to classify a wide range of biological diversity or if you want to classify sequences from unknown sources. However, if you are only interested in specific species or samples or do not require highly accurate classification results, it may not be necessary to include the NT database.\n",
    "\n",
    "It is important to note that including the NT database in the Kraken database will increase the size and build time of the database, so specific needs should be considered when making this decision."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## false positive in database\n",
    "![](https://journals.plos.org/ploscompbiol/article/figure/image?size=large&id=10.1371/journal.pcbi.1006277.g005)\n",
    ">Top 10 species identified in corneal samples per database.\n",
    ">\n",
    ">The non-human reads from the 20 corneal samples were classified against four different Kraken databases: the original EuPathDB (A), EuPathDB-clean (B), RefSeq EuPathDB (C), and the final MicrobeDB (D). The plot above shows the 10 species with the most classified reads per megabase in a single corneal sample.{cite:p}`lu2018removing`\n",
    ">\n",
    ">https://doi.org/10.1371/journal.pcbi.1006277.g005"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, plasmid sequences are often included in bacterial genome sequence data submitted to NCBI, which can lead to incorrect taxonomic classification. After removing all plasmid sequences included in bacterial RefSeq genomes and reassigning them to a distinct taxonomic group, the proportion of reads correctly classified to a specific bacterium was reduced {cite:p}`doster_cautionary_2019`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kraken2-inspect /data/database/RefSeqV205_500G  | head -5"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "K-mer-based sequence classifiers such as Kraken2 and KrakenUniq can often make erroneous assignments due to genome similarities, such as the false positive phenomenon of assigning human reads to Clostridium botulinum, which can occur.(See [Kraken2 issue #621:Host Reads Being Classified as MIcrobial reads](https://github.com/DerrickWood/kraken2/issues/621))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Daabase contruct\n",
    "since kraken2 origin database constructed is so poor.we use manual script to download and build it.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python /data/project/host-microbiome/kraken_metaphlan_comparison/database_building_scripts/get_ncbi_other_domains.py --domain fungi,viral,vertebrate_mammalian,bacteria,archaea --complete True --folder  /data/database/kraken2_RefSeqV217_Complete_Chrom/download/ --download_genomes True --log_file ./run_all_download.log --processors 50"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then in the path you will get summary_to_download.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44588\n"
     ]
    }
   ],
   "source": [
    "!grep -c \"Complete Genome\" /data/database/kraken2_RefSeqV217_Complete_Chrom/download/summary_to_download.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5394\n"
     ]
    }
   ],
   "source": [
    "!grep -c \"Chromosome\" /data/database/kraken2_RefSeqV217_Complete_Chrom/download/summary_to_download.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kraken2-build --download-taxonomy --db kraken2_RefSeqV217_Complete_Chrom --use-ftp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/database/kraken2_RefSeqV217_Complete_Chrom/download\n"
     ]
    }
   ],
   "source": [
    "%cd /data/database/kraken2_RefSeqV217_Complete_Chrom/download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>version_status</th>\n",
       "      <th>...</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th># assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_002287175.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04229035</td>\n",
       "      <td>LMVM00000000.1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>2161</td>\n",
       "      <td>2161</td>\n",
       "      <td>Methanobacterium bryantii</td>\n",
       "      <td>strain=M.o.H.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2017/09/06</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>University of California Santa Barbara</td>\n",
       "      <td>GCA_002287175.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>assembly from type material</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000762265.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03085433</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>2162</td>\n",
       "      <td>2162</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>strain=BRM9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2014/10/02</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>PGgRc</td>\n",
       "      <td>GCA_000762265.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001458655.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA2779801</td>\n",
       "      <td>NaN</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>2162</td>\n",
       "      <td>2162</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mb9</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2015/11/16</td>\n",
       "      <td>Mb9</td>\n",
       "      <td>CEBITEC</td>\n",
       "      <td>GCA_001458655.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015351695.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16521386</td>\n",
       "      <td>JADIIL000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2162</td>\n",
       "      <td>2162</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bin2</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2020/11/10</td>\n",
       "      <td>ASM1535169v1</td>\n",
       "      <td>Water Research Institute, IRSA-CNR</td>\n",
       "      <td>GCA_015351695.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>derived from metagenome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000824705.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA2796325</td>\n",
       "      <td>CCXV00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2173</td>\n",
       "      <td>2173</td>\n",
       "      <td>Methanobrevibacter smithii</td>\n",
       "      <td>strain=ACE6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2014/10/02</td>\n",
       "      <td>Methanobrevibacter smithii</td>\n",
       "      <td>URMITE</td>\n",
       "      <td>GCA_000824705.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029338335.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN33708874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>3034020</td>\n",
       "      <td>3034020</td>\n",
       "      <td>Halovivax sp. TS33</td>\n",
       "      <td>strain=TS33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2023/03/22</td>\n",
       "      <td>ASM2933833v1</td>\n",
       "      <td>Jiangsu University</td>\n",
       "      <td>GCA_029338335.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029338355.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN33716776</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>3034023</td>\n",
       "      <td>3034023</td>\n",
       "      <td>Halosegnis sp. DT85</td>\n",
       "      <td>strain=DT85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2023/03/22</td>\n",
       "      <td>ASM2933835v1</td>\n",
       "      <td>Jiangsu University</td>\n",
       "      <td>GCA_029338355.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029338375.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN33716851</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>3034024</td>\n",
       "      <td>3034024</td>\n",
       "      <td>Halorussus sp. DT80</td>\n",
       "      <td>strain=DT80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2023/03/22</td>\n",
       "      <td>ASM2933837v1</td>\n",
       "      <td>Jiangsu University</td>\n",
       "      <td>GCA_029338375.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029338395.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN33716999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>3034025</td>\n",
       "      <td>3034025</td>\n",
       "      <td>Halorussus sp. DT72</td>\n",
       "      <td>strain=DT72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2023/03/22</td>\n",
       "      <td>ASM2933839v1</td>\n",
       "      <td>Jiangsu University</td>\n",
       "      <td>GCA_029338395.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029338515.1</th>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN33717057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>3034026</td>\n",
       "      <td>3034026</td>\n",
       "      <td>Haloglomus sp. DT116</td>\n",
       "      <td>strain=DT116</td>\n",
       "      <td>NaN</td>\n",
       "      <td>latest</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2023/03/22</td>\n",
       "      <td>ASM2933851v1</td>\n",
       "      <td>Jiangsu University</td>\n",
       "      <td>GCA_029338515.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1458 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       bioproject     biosample         wgs_master  \\\n",
       "# assembly_accession                                                 \n",
       "GCF_002287175.1       PRJNA224116  SAMN04229035     LMVM00000000.1   \n",
       "GCF_000762265.1       PRJNA224116  SAMN03085433                NaN   \n",
       "GCF_001458655.1       PRJNA224116  SAMEA2779801                NaN   \n",
       "GCF_015351695.1       PRJNA224116  SAMN16521386  JADIIL000000000.1   \n",
       "GCF_000824705.1       PRJNA224116  SAMEA2796325     CCXV00000000.1   \n",
       "...                           ...           ...                ...   \n",
       "GCF_029338335.1       PRJNA224116  SAMN33708874                NaN   \n",
       "GCF_029338355.1       PRJNA224116  SAMN33716776                NaN   \n",
       "GCF_029338375.1       PRJNA224116  SAMN33716851                NaN   \n",
       "GCF_029338395.1       PRJNA224116  SAMN33716999                NaN   \n",
       "GCF_029338515.1       PRJNA224116  SAMN33717057                NaN   \n",
       "\n",
       "                            refseq_category    taxid  species_taxid  \\\n",
       "# assembly_accession                                                  \n",
       "GCF_002287175.1       representative genome     2161           2161   \n",
       "GCF_000762265.1                          na     2162           2162   \n",
       "GCF_001458655.1       representative genome     2162           2162   \n",
       "GCF_015351695.1                          na     2162           2162   \n",
       "GCF_000824705.1                          na     2173           2173   \n",
       "...                                     ...      ...            ...   \n",
       "GCF_029338335.1                          na  3034020        3034020   \n",
       "GCF_029338355.1                          na  3034023        3034023   \n",
       "GCF_029338375.1                          na  3034024        3034024   \n",
       "GCF_029338395.1                          na  3034025        3034025   \n",
       "GCF_029338515.1                          na  3034026        3034026   \n",
       "\n",
       "                                    organism_name infraspecific_name isolate  \\\n",
       "# assembly_accession                                                           \n",
       "GCF_002287175.1         Methanobacterium bryantii      strain=M.o.H.     NaN   \n",
       "GCF_000762265.1       Methanobacterium formicicum        strain=BRM9     NaN   \n",
       "GCF_001458655.1       Methanobacterium formicicum                NaN     Mb9   \n",
       "GCF_015351695.1       Methanobacterium formicicum                NaN    bin2   \n",
       "GCF_000824705.1        Methanobrevibacter smithii        strain=ACE6     NaN   \n",
       "...                                           ...                ...     ...   \n",
       "GCF_029338335.1                Halovivax sp. TS33        strain=TS33     NaN   \n",
       "GCF_029338355.1               Halosegnis sp. DT85        strain=DT85     NaN   \n",
       "GCF_029338375.1               Halorussus sp. DT80        strain=DT80     NaN   \n",
       "GCF_029338395.1               Halorussus sp. DT72        strain=DT72     NaN   \n",
       "GCF_029338515.1              Haloglomus sp. DT116       strain=DT116     NaN   \n",
       "\n",
       "                     version_status  ... genome_rep seq_rel_date  \\\n",
       "# assembly_accession                 ...                           \n",
       "GCF_002287175.1              latest  ...       Full   2017/09/06   \n",
       "GCF_000762265.1              latest  ...       Full   2014/10/02   \n",
       "GCF_001458655.1              latest  ...       Full   2015/11/16   \n",
       "GCF_015351695.1              latest  ...       Full   2020/11/10   \n",
       "GCF_000824705.1              latest  ...       Full   2014/10/02   \n",
       "...                             ...  ...        ...          ...   \n",
       "GCF_029338335.1              latest  ...       Full   2023/03/22   \n",
       "GCF_029338355.1              latest  ...       Full   2023/03/22   \n",
       "GCF_029338375.1              latest  ...       Full   2023/03/22   \n",
       "GCF_029338395.1              latest  ...       Full   2023/03/22   \n",
       "GCF_029338515.1              latest  ...       Full   2023/03/22   \n",
       "\n",
       "                                        asm_name  \\\n",
       "# assembly_accession                               \n",
       "GCF_002287175.1                      ASM228717v1   \n",
       "GCF_000762265.1                       ASM76226v1   \n",
       "GCF_001458655.1                              Mb9   \n",
       "GCF_015351695.1                     ASM1535169v1   \n",
       "GCF_000824705.1       Methanobrevibacter smithii   \n",
       "...                                          ...   \n",
       "GCF_029338335.1                     ASM2933833v1   \n",
       "GCF_029338355.1                     ASM2933835v1   \n",
       "GCF_029338375.1                     ASM2933837v1   \n",
       "GCF_029338395.1                     ASM2933839v1   \n",
       "GCF_029338515.1                     ASM2933851v1   \n",
       "\n",
       "                                                   submitter  gbrs_paired_asm  \\\n",
       "# assembly_accession                                                            \n",
       "GCF_002287175.1       University of California Santa Barbara  GCA_002287175.1   \n",
       "GCF_000762265.1                                        PGgRc  GCA_000762265.1   \n",
       "GCF_001458655.1                                      CEBITEC  GCA_001458655.1   \n",
       "GCF_015351695.1           Water Research Institute, IRSA-CNR  GCA_015351695.1   \n",
       "GCF_000824705.1                                       URMITE  GCA_000824705.1   \n",
       "...                                                      ...              ...   \n",
       "GCF_029338335.1                           Jiangsu University  GCA_029338335.1   \n",
       "GCF_029338355.1                           Jiangsu University  GCA_029338355.1   \n",
       "GCF_029338375.1                           Jiangsu University  GCA_029338375.1   \n",
       "GCF_029338395.1                           Jiangsu University  GCA_029338395.1   \n",
       "GCF_029338515.1                           Jiangsu University  GCA_029338515.1   \n",
       "\n",
       "                     paired_asm_comp  \\\n",
       "# assembly_accession                   \n",
       "GCF_002287175.1            identical   \n",
       "GCF_000762265.1            identical   \n",
       "GCF_001458655.1            identical   \n",
       "GCF_015351695.1            identical   \n",
       "GCF_000824705.1            identical   \n",
       "...                              ...   \n",
       "GCF_029338335.1            identical   \n",
       "GCF_029338355.1            identical   \n",
       "GCF_029338375.1            identical   \n",
       "GCF_029338395.1            identical   \n",
       "GCF_029338515.1            identical   \n",
       "\n",
       "                                                               ftp_path  \\\n",
       "# assembly_accession                                                      \n",
       "GCF_002287175.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_000762265.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_001458655.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_015351695.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_000824705.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "...                                                                 ...   \n",
       "GCF_029338335.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_029338355.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_029338375.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_029338395.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "GCF_029338515.1       https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "                         excluded_from_refseq    relation_to_type_material  \\\n",
       "# assembly_accession                                                         \n",
       "GCF_002287175.1                           NaN  assembly from type material   \n",
       "GCF_000762265.1                           NaN                          NaN   \n",
       "GCF_001458655.1                           NaN                          NaN   \n",
       "GCF_015351695.1       derived from metagenome                          NaN   \n",
       "GCF_000824705.1                           NaN                          NaN   \n",
       "...                                       ...                          ...   \n",
       "GCF_029338335.1                           NaN                          NaN   \n",
       "GCF_029338355.1                           NaN                          NaN   \n",
       "GCF_029338375.1                           NaN                          NaN   \n",
       "GCF_029338395.1                           NaN                          NaN   \n",
       "GCF_029338515.1                           NaN                          NaN   \n",
       "\n",
       "                     asm_not_live_date  \n",
       "# assembly_accession                    \n",
       "GCF_002287175.1                     na  \n",
       "GCF_000762265.1                     na  \n",
       "GCF_001458655.1                     na  \n",
       "GCF_015351695.1                     na  \n",
       "GCF_000824705.1                     na  \n",
       "...                                ...  \n",
       "GCF_029338335.1                     na  \n",
       "GCF_029338355.1                     na  \n",
       "GCF_029338375.1                     na  \n",
       "GCF_029338395.1                     na  \n",
       "GCF_029338515.1                     na  \n",
       "\n",
       "[1458 rows x 22 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly = pd.read_csv('archaea_assembly_summary.txt', header=1, index_col=0,sep=\"\\t\")\n",
    "assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2240398/129226915.py:6: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  assembly = pd.read_csv(assembly, header=1, index_col=0,sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "assembly_lists = ['archaea_assembly_summary.txt', 'bacteria_assembly_summary.txt', 'fungi_assembly_summary.txt', 'vertebrate_mammalian_assembly_summary.txt', 'viral_assembly_summary.txt']\n",
    "genome_taxid = []\n",
    "for assembly in assembly_lists:\n",
    "  assembly = pd.read_csv(assembly, header=1, index_col=0,sep=\"\\t\")\n",
    "  for row in assembly.index.values:\n",
    "    genome_taxid.append([row, assembly.loc[row, 'taxid']])\n",
    "\n",
    "genome_taxid = pd.DataFrame(genome_taxid, columns=['Genome accession', 'taxid']).set_index('Genome accession')\n",
    "genome_taxid.to_csv('db_samples.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_002287175.1</th>\n",
       "      <td>2161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000762265.1</th>\n",
       "      <td>2162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001458655.1</th>\n",
       "      <td>2162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015351695.1</th>\n",
       "      <td>2162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000824705.1</th>\n",
       "      <td>2173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_027946335.1</th>\n",
       "      <td>3003632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_027946345.1</th>\n",
       "      <td>3003632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_027574445.1</th>\n",
       "      <td>3003729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_028515065.1</th>\n",
       "      <td>3020045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_028674785.1</th>\n",
       "      <td>3025408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>295968 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    taxid\n",
       "Genome accession         \n",
       "GCF_002287175.1      2161\n",
       "GCF_000762265.1      2162\n",
       "GCF_001458655.1      2162\n",
       "GCF_015351695.1      2162\n",
       "GCF_000824705.1      2173\n",
       "...                   ...\n",
       "GCF_027946335.1   3003632\n",
       "GCF_027946345.1   3003632\n",
       "GCF_027574445.1   3003729\n",
       "GCF_028515065.1   3020045\n",
       "GCF_028674785.1   3025408\n",
       "\n",
       "[295968 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "python /data/project/host-microbiome/kraken_metaphlan_comparison/database_building_scripts/rename_fasta_headers.py --genome_folder /data/database/kraken2_RefSeqV217_Complete_Chrom/download --genome_list /data/database/kraken2_RefSeqV217_Complete_Chrom/download/db_samples.tsv --log_file logfile.txt --processors 40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "python /data/project/host-microbiome/kraken_metaphlan_comparison/database_building_scripts/unzip_add_library.py --genome_folder fasta_renamed_RefSeqV205_Complete --database RefSeqV205_Complete --processors 12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```{bibliography}\n",
    ":style: unsrt\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 (conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "15cbaadd938628f9eba240e0048828d8c3674287d0229126efcaad1887a1387e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
